cpp/src/arrow/table.cc

```diff
@@ -601,6 +602,19 @@ TableBatchReader::TableBatchReader(const Table& table)
   }
 }
 
+TableBatchReader::TableBatchReader(std::shared_ptr<Table> table)
+    : table_(*table.get()),
```

Comment:
nit: would it work better to just call the other constructor instead of repeating it here, e.g. `... : TableBatchReader(*table.get()), owned_table_(std::move(table)) {}`

Reply:
I thought of following this approach at first, but an error will be thrown in this case, as an initializer for a delegating constructor must appear alone

Reply:
Huh, forgot about that. Oh well, it's only a short initializer.